Core/GDCore/IDE/Events/ProjectElementRenamer.cpp

```diff
@@ -0,0 +1,219 @@
+          occurrences.push_back(parameterNode->location);
+        } else {
+          parameterNode->Visit(*this);
+        }
+      }
+    }
+  }
+  void OnVisitEmptyNode(EmptyNode &node) override {}
+
+private:
+  const gd::Platform &platform;
+  const gd::ObjectsContainer &globalObjectsContainer;
+  const gd::ObjectsContainer &objectsContainer;
+  const gd::String &expressionPlainString;
+  const gd::String &oldName;
+  const gd::String parameterType;
+  /// If not empty, parameters will be taken into account only if related to
+  /// this object.
+  const gd::String objectName;
+  /// If not empty, parameters will be taken into account only if related to
+  /// this layer.
+  const gd::String layerName;
+  std::vector<gd::ExpressionParserLocation> occurrences;
+};
+
+bool ProjectElementRenamer::DoVisitInstruction(gd::Instruction &instruction,
+                                               bool isCondition) {
+  const auto &metadata = isCondition
+                             ? gd::MetadataProvider::GetConditionMetadata(
+                                   platform, instruction.GetType())
+                             : gd::MetadataProvider::GetActionMetadata(
+                                   platform, instruction.GetType());
+
+  gd::String lastLayerName;
+  gd::ParameterMetadataTools::IterateOverParametersWithIndex(
+      instruction.GetParameters(), metadata.GetParameters(),
+      [&](const gd::ParameterMetadata &parameterMetadata,
+          const gd::Expression &parameterValue, size_t parameterIndex,
+          const gd::String &lastObjectName) {
+        if (parameterMetadata.GetType() == "layer") {
+          // Remove quotes, it won't match if it's not a literal anyway.
+          lastLayerName = parameterValue.GetPlainString().substr(
+              1, parameterValue.GetPlainString().length() - 2);
+        }
+
+        if (parameterMetadata.GetType() == parameterType &&
+            (objectName.empty() || lastObjectName == objectName) &&
+            (layerName.empty() || lastLayerName == layerName)) {
+          if (parameterValue.GetPlainString() == "\"" + oldName + "\"") {
+            instruction.SetParameter(parameterIndex,
+                                     gd::Expression("\"" + newName + "\""));
+          }
+        }
+        auto node = parameterValue.GetRootNode();
+        if (node) {
+          ExpressionIdentifierStringFinder finder(
+              platform, GetGlobalObjectsContainer(), GetObjectsContainer(),
+              parameterValue.GetPlainString(), parameterType, objectName,
+              layerName, oldName);
+          node->Visit(finder);
+
+          if (finder.GetOccurrences().size() > 0) {
+
+            gd::String newNameWithQuotes = "\"" + newName + "\"";
+            gd::String oldParameterValue = parameterValue.GetPlainString();
+            gd::String newParameterValue;
+            auto previousEndPosition = 0;
+            for (auto &&occurrenceLocation : finder.GetOccurrences()) {
```

Comment:
Like, will this make locations invalid once one occurrence is replaced?

Reply:
I think it should be fine.
The original string of the Expression is untouched and a new Expression is instantiated. The parser will be called when the node is asked to the Expression.
```Cpp
instruction.SetParameter(parameterIndex, gd::Expression(newParameterValue));
```

Reply:
Alright, double check with something like: `CameraCenterX("my layer") + CameraCenterX("my layer")` to check that it works for both a longer name or a shorter name when renamed:
`CameraCenterX("my layer 2") + CameraCenterX("my layer 2")`
`CameraCenterX("mylayer") + CameraCenterX("mylayer")`

Reply:
Actually I see you have a test for this already :) "Can update layer names in events"
Can you do this with a shorter name too?

I'm worried because I see we're doing changes by string manipulation and not by modifying the nodes and printing them back to a string with `ExpressionParser2NodePrinter`.

Reply:
> Actually I see you have a test for this already :) "Can update layer names in events" Can you do this with a shorter name too?
> 
> I'm worried because I see we're doing changes by string manipulation and not by modifying the nodes and printing them back to a string with `ExpressionParser2NodePrinter`.

I will add it. It's a valid test case in case the implementation changes. Even though, there is no string manipulation done, it creates a new string so it won't cover anything.